embeddings_utils.go

```diff
@@ -0,0 +1,31 @@
+package openai
+
+import "math"
+
+// CosineSimilarity Calculate cosine similarity.
+// Note:
+// We recommend cosine similarity. The choice of distance function typically doesn’t matter much.
+// OpenAI embeddings are normalized to length 1, which means that:
+// Cosine similarity can be computed slightly faster using just a dot product.
+// Cosine similarity and Euclidean distance will result in the identical rankings.
+// See: https://platform.openai.com/docs/guides/embeddings/limitations-risks
+func CosineSimilarity(v1, v2 []float32) float32 {
+	// Calculate dot product
+	dot := DotProduct(v1, v2)
+	// Calculate magnitude of v1
+	v1Magnitude := math.Sqrt(float64(DotProduct(v1, v1)))
```

Comment:
Embeddings are normalized to length 1, so we don't need to do that. CosineSimilarity is equal to DotProduct in this case.

https://platform.openai.com/docs/guides/embeddings/which-distance-function-should-i-use

Reply:
@sashabaranov  OK I just delete the function CosineSimilarity and it’s over. Anyway, this is not used in the example.